### DISCUSSION TITLE
Can't create 3rd party CRD

### DISCUSSION THREAD

**Discussion by benkeil** [kube-rs/kube#1296] [Q&A] [answered]
I made my own struct to representing the 3rd party CRD:

```rust
#[derive(Clone, Serialize, Deserialize, Debug)]
#[serde(rename_all = "camelCase")]
pub struct TerraformSpec {
    pub terraform_version: String,
    pub terraform_module: TerraformModule,
    pub backend: String,
    pub ignore_delete: bool,
    pub task_options: Vec<TaskOption>,
}

pub type Terraform = Object<TerraformSpec, NotUsed>;
```

Initialising the API client:

```rust
let gvk = GroupVersionKind::gvk(&"tf.galleybytes.com", &"v1beta1", &"Terraform");
let (ar, caps) = kube::discovery::pinned_kind(&ctx.client.clone(), &gvk)
    .await
    .expect(format!("could not get gvk {:?}", gvk).as_str());
let terraform_api = Api::<Terraform>::all_with(ctx.client.clone(), &ar);
```

Some information about the API:

```
api resource: ApiResource { group: "tf.galleybytes.com", version: "v1beta1", api_version: "tf.galleybytes.com/v1beta1", kind: "Terraform", plural: "terraforms" }
supported operations: ["delete", "deletecollection", "get", "list", "patch", "create", "update", "watch"]
supports create: true
```

List works and returns items:

```rust
let list = terraform_api.list(&Default::default()).await.unwrap();
```

But when I try to Crete a new resource I get a 404:

```rust
let resource = Terraform::new(
    name.as_str(),
    &ar,
    TerraformSpec {
        backend: format!(
            r#"
            terraform {{
              backend "kubernetes" {{
                secret_suffix    = "{name}"
                in_cluster_config  = true
                namespace = "{namespace}"
              }}
            }}
            "#
        ),
        ignore_delete: false,
        terraform_module: TerraformModule {
            source: "terraform-aws-modules/sqs/aws?ref=4.0.2".into(),
        },
        terraform_version: "1.5.4".into(),
        task_options: vec![TaskOption {
            r#for: vec!["*".into()],
            env: vec![Env::tf_var("name".into(), "foobar".into())],
        }],
    },
);

let created = terraform_api
    .create(&PostParams::default(), &resource)
    .await;
```

```
Err(Api(ErrorResponse { status: "404 Not Found", message: "\"404 page not found\\n\"", reason: "Failed to parse error data", code: 404 })))
```

**Comment by clux** [comment, marked as answer]
404 makes me think that maybe you are using `Api::all_with` on a namespaced resource, but that's just a hunch. Is that what you are doing? If so try `Api::namespaced_with`. See https://github.com/kube-rs/kube/issues/1276 for some context of why you need both an `Api::namespaced` and an `Api::all` atm if you want to do both global lists + namespaced create/get.

Btw, I see you are using `Object` and a dynamic api on top even though you have typed structs. You don't need to use the dynamic api for this. You could import `kube::CustomResource` with the `derive` feature and add

```rust
#[derive(CustomResource)]
#[kube(group = "tf.galleybytes.com" version = "v1beta1")]
#[kube(namespaced)] // if you are namespaced
struct TerraformSpec { ..as_before() }
```

so that you can use the fully typed api:

```rust
let tfs_all: Api<Terraform> = Api::all(client.clone()); // for --all style listing
let tfs_ns: Api<Terraform> = Api::default_namespaced(client.clone()); // for namespaced .get or .create
```

without needing to do api discovery

**Comment by benkeil** [comment]
I didn't used the `CustomResource` `derive` because it's not my resource. So you would still recommend to use it?

`Api::namespaced_with` works, thank you!!

What is the equivalent of `kubectl apply`? Is it a patch operation?

**Reply by clux** [reply]
Yeah, i personally use `CustomResource` on everything because it's easier and safer. It's also pretty easy to integrate from third-party crds with something like [kopium](https://github.com/kube-rs/kopium) where the struct and derive will be generated for you.

**Reply by clux** [reply]
`kubectl apply` is basically `api.patch` yes. replicating `--server-side` needs the server-side `PatchParams`, client-side needs one of the older styles of patch parameters. there's more docs in the examples and on the `Api` type in docs.rs/kube

**Reply by benkeil** [reply]
I understand everything except this: `let params = PatchParams::apply("myapp");` what does `myapp` reference to?

**Reply by clux** [reply]
it's just the name of the field-manager (the controller in charge of updating the fields). it's required to have one for serverside apply, as different controllers (and users) can update different parts of the object. you can see who owns what part of the schema with `kubectl get deploys --show-managed-fields` say

**Reply by benkeil** [reply]
Again, thank you very much for your support ☺️
